monai/utils/dist.py

```diff
@@ -0,0 +1,151 @@
+        concat: whether to concat the gathered list to be a Tensor, if False, return a list
+            of Tensors, similar behavior as torch.distributed.all_gather(). default to True.
+
+    Note:
+        The input data on different ranks must have exactly same `dtype`.
+
+    """
+    if not isinstance(data, torch.Tensor):
+        raise ValueError("input data must be PyTorch Tensor.")
+    # data of all the ranks must have same number of dimensions
+    ndims = data.ndimension()
+    length: int = data.shape[0] if ndims > 0 else 1
+
+    def _torch_all_gather(data: torch.Tensor) -> List[torch.Tensor]:
+        """
+        Implementation based on native PyTorch distributed data parallel APIs.
+
+        """
+        device = get_dist_device()
+        orig_device = data.device
+        data = data.to(device)
+        data = data.unsqueeze(0) if ndims == 0 else data
+
+        # make sure the data is evenly-divisible on multi-GPUs
+        length_tensor = torch.as_tensor([length], device=device)
+        all_lens = [torch.zeros_like(length_tensor) for _ in range(dist.get_world_size())]
+        dist.all_gather(all_lens, length_tensor)
+        all_lens_: List[int] = [int(i.item()) for i in all_lens]
+
+        max_len: int = max(all_lens_)
+        if length < max_len:
+            size = [max_len - length] + list(data.shape[1:])
+            data = torch.cat([data, data.new_full(size, 0)], dim=0)
+        # all gather across all processes
+        output = [torch.zeros_like(data) for _ in range(dist.get_world_size())]
+        dist.all_gather(output, data)
+        # remove the padding items, if all the input data doesn't have batch dim, suqeeze the first dim
+        return [(o.squeeze(0) if ndims == 0 else o[:l, ...]).to(orig_device) for o, l in zip(output, all_lens_)]
+
+    def _ignite_all_gather(data: torch.Tensor) -> List[torch.Tensor]:
+        """
+        Implementation based on PyTorch ignite package, it can support more kinds of backends.
+
+        """
+        data = data.unsqueeze(0) if ndims == 0 else data
+        # make sure the data is evenly-divisible on multi-GPUs
+        all_lens: List[int] = idist.all_gather(length)
+        max_len: int = max(all_lens)
+        if length < max_len:
+            size = [max_len - length] + list(data.shape[1:])
+            data = torch.cat([data, data.new_full(size, 0)], dim=0)
+        # all gather across all processes
+        output = idist.all_gather(data)
+        # delete the padding NaN items
+        if ndims == 0:
+            # if all the input data doesn't have batch dim, unbind to a list of 0-dim Tensors
+            return list(torch.unbind(output, dim=0))
+        return [output[i * max_len : i * max_len + l, ...] for i, l in enumerate(all_lens)]
+
+    output: List[torch.Tensor]
+    if dist.is_available() and dist.is_initialized():
+        if dist.get_world_size() <= 1:
+            return data
+        output = _torch_all_gather(data=data)
+    elif has_ignite:
```

Comment:
I think we should prefer ignite if it's installed otherwise fall back to torch native?

Reply:
Hi @wyli , thanks for your suggestion.
The issue is that: there are some cases that if you don't init your dist program with ignite APIs, the ignite dist calls will not work, for example: https://github.com/pytorch/ignite/issues/2048#issuecomment-858331410.
But PyTorch dist APIs can work if initialized with ignite APIs.
So I think we should use PyTorch dist APIs by default, if users want to use ignite implementation, they should totally use ignite dist APIs in their program.
What do you think?

Thanks.

Reply:
I'm not sure about the idea here. @vfdev-5 could you please help confirm it?

Reply:
Hi @Nic-Ma @wyli , the issue pytorch/ignite#2048 revealed an issue we have in the way how ignite DiskSaver class interacts with distributed group created by user using pytorch in a single rank. In few words the issue is the following:
```python
dist.init_process_group(...)

# ignite is not aware of created DDP group

if dist.get_rank() == 0:
    # - Here only 1 process passes
    
    # - Now user executes something that calls ignite.distributed the first time 
    # where ignite internally calls a collective op => execution hangs
    
    # - To avoid that we propose to call `idist.sync()` right after `dist.init_process_group(...)`

dist.destroy_process_group()
```

> The issue is that: there are some cases that if you don't init your dist program with ignite APIs, the ignite dist calls will not work

If this method is called by all processes there wont be any issue. And according to the name `evenly_divisible_all_gather` it will be called by all DDP processes...

Reply:
Hi @vfdev-5 , thanks for your explanation!
So I think you mean we can change as @wyli suggested? Prefer ignite, if not installed, fall back to torch native?

Thanks.

Reply:
For me both variants should work here and I have no particular preference for any. But, yes, we can change as @wyli suggested.

Thanks

Reply:
Hi @wyli , I have changed the PR to prefer to ignite implementation.
Could you please help review it again?

Thanks in advance.